src/lib/storage/table.cpp

```diff
@@ -100,17 +102,54 @@ bool Table::column_nullable(const ColumnID column_id) const {
 
 std::shared_ptr<Chunk> Table::get_chunk(ChunkID chunk_id) {
   Assert(chunk_id < chunk_count(), "Chunk with ID does not exist.");
+  // We cannot hand out this chunk if compression is currently in progress, as somebody might modify it,
+  // causing us to lose this inserted data once the compression (which was based on the old chunk state) has finished.
+  // If the caller can ensure he won't modify the chunk, he can use the const overload instead.
+  const std::lock_guard<std::recursive_mutex> lock(_chunk_access_lock);
   return _chunks[chunk_id];
 }
 
 std::shared_ptr<const Chunk> Table::get_chunk(ChunkID chunk_id) const {
   Assert(chunk_id < chunk_count(), "Chunk with ID does not exist.");
+  // We can omit locking our _chunk_access_lock here because we can be sure that the chunk we hand out will not be
+  // modified (because it is const).
+  // This means getting a chunk that way is safe even if we have a compression currently in progress on that chunk,
+  // because we don't have to worry that we lose some data that gets inserted after the start of the compression.
   return _chunks[chunk_id];
 }
 
+void Table::compress_segment_and_add_to_chunk(ColumnID index, const std::shared_ptr<Chunk>& new_chunk,
+                                              const std::shared_ptr<Chunk>& chunk_to_be_compressed) {
+  const auto segment = chunk_to_be_compressed->get_segment(index);
+  resolve_data_type(column_type(index), [&new_chunk, &segment](const auto data_type_t) {
+    using ColumnDataType = typename decltype(data_type_t)::type;
+    new_chunk->add_segment(std::make_shared<DictionarySegment<ColumnDataType>>(segment));
```

Comment:
Possible race condition between segment ordering: 
Using `new_chunk->add_segment()` blindly may insert a segment out-of-order if its thread finished faster than another thread. Consider passing a segment index to each thread and writing into a vector and then finally merging that vector into a new chunk.
Also: Your implementation of `Chunk::add_segment()` uses `vector::emplace_back` which is [not thread-safe](https://en.cppreference.com/w/cpp/container#Thread_safety).